File: fe/js/GamePage/Ball.js
class Ball {
	// constructor(angle, speed, radius) {
	// 	this.angle = angle;
	// 	this.speed = speed;
	// 	this.radius = radius;
	// }
	constructor(speed, radius) {
		this.speed = speed;
		this.radius = radius;
	}

	// initBall(boardWidth, boardHeight) {
	// 	this.xPos = boardWidth / 2;
	// 	this.yPos = boardHeight / 2;
	// 	const dir = this._calculateBallDirection();
	// 	this.dx = dir.dx;
	// 	this.dy = dir.dy;
	// 	this.angle = 0;
	// }
	initBall(x, y, angle) {
		this.xPos = x;
		this.yPos = y;
		this.angle = angle;
		const dir = this._calculateBallDirection();
		this.dx = dir.dx;
		this.dy = dir.dy;
	}

	_calculateBallDirection() {
		const angleRadians = (this.angle * Math.PI) / 180;
		const dx = Math.cos(angleRadians) * this.speed;
		const dy = Math.sin(angleRadians) * this.speed;
		return { dx, dy };
	}

	reversalRandomDx() {
		let rand = Math.floor(Math.random() * 81) - 40; // -40 ~ +40도 사이에서 이동 방향 변화
		this.angle = Math.max(0, Math.min(45, this.angle + rand)); // 최소 각도 0, 최대 각도 45
		const dir = this._calculateBallDirection(this.speed, this.angle);
		this.dx = this.dx < 0 ? dir.dx : -dir.dx; // dx는 부호 반전
		this.dy = this.dy < 0 ? -dir.dy : dir.dy; // dy는 부호 유지
	}

	getRightX() {
		return this.xPos + this.radius;
	}

	getLeftX() {
		return this.xPos - this.radius;
	}

	getTopY() {
		return this.yPos - this.radius;
	}

	getBottomY() {
		return this.yPos + this.radius;
	}
}

export default Ball;

File: BE/pingpong/events/ball.py
import math
import random

class Ball:
    def __init__(self, speed, radius):
        self.speed = speed
        self.radius = radius
        self.xPos = 0
        self.yPos = 0
        self.angle = 0
        self.dx = 0
        self.dy = 0

    def reset_ball(self, x, y, angle):
        self.xPos = x
        self.yPos = y
        self.angle = angle
        dir = self._calculate_ball_direction()
        self.dx = dir['dx']
        self.dy = dir['dy']

    def move(self):
        self.xPos += self.dx
        self.yPos += self.dy

    def _calculate_ball_direction(self):
        angle_radians = (self.angle * math.pi) / 180
        dx = math.cos(angle_radians) * self.speed
        dy = math.sin(angle_radians) * self.speed
        return {'dx': dx, 'dy': dy}

    def reversal_random_dx(self):
        rand = random.randint(-40, 40)  # -40 ~ +40도 사이에서 이동 방향 변화
        self.angle = max(0, min(45, self.angle + rand))  # 최소 각도 0, 최대 각도 45
        dir = self._calculate_ball_direction()
        self.dx = dir['dx'] if self.dx < 0 else -dir['dx']  # dx는 부호 반전
        self.dy = -dir['dy'] if self.dy < 0 else dir['dy']  # dy는 부호 유지

    def change_direction(self, angle):
        self.angle = angle
        dir = self._calculate_ball_direction()
        self.dx = dir['dx']
        self.dy = dir['dy']

    def get_right_x(self):
        return self.xPos + self.radius

    def get_left_x(self):
        return self.xPos - self.radius

    def get_top_y(self):
        return self.yPos - self.radius

    def get_bottom_y(self):
        return self.yPos + self.radius
